plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSink.java

```diff
@@ -564,14 +630,26 @@ public int[] partitionPage(Page page)
             Block[] blocks = new Block[columns.size()];
             for (int i = 0; i < columns.size(); i++) {
                 PartitionColumn column = columns.get(i);
-                Block block = page.getBlock(column.getSourceChannel());
+                Block block = getPartitionBlock(column, page);
                 blocks[i] = column.getBlockTransform().apply(block);
             }
             Page transformed = new Page(page.getPositionCount(), blocks);
 
             return pageIndexer.indexPage(transformed);
         }
 
+        private Block getPartitionBlock(PartitionColumn column, Page page)
+        {
+            ListIterator<Integer> iterator = column.getSourceChannels().listIterator();
+            Block block = page.getBlock(iterator.next());
+
+            while (iterator.hasNext()) {
+                block = toColumnarRow(block).getField(iterator.next());
+            }
```

Comment:
Typically we try to avoid using streams or iterators in the code to process a Page and use simple for loops for performance reasons.